File: frontend/src/lib/chat/application-commands.test.ts
import { describe, expect, it } from 'vitest';
import {
  commandAttachmentOptionIds,
  applicationCommandAllowedByChannelPermissions,
  applicationIntegrationAllowedByUsePermission,
  applicationCommandAllowedByUsePermission,
  applicationCommandRequestIdentity,
  applicationCommandByIdentity,
  applicationCommandLauncherGroups,
  commandCompletions,
  commandComposerOptions,
  commandInvocation,
  commandOptionAllowsChannelType,
  commandOptionPayload,
  commandOptionsComplete,
  commandStringOptions,
  localizedCommandText,
  parseApplicationCommandAutocompleteChoices,
  parseApplicationCommands,
  resolveCommandInvocation,
  uniqueChatInputCommand,
  type ApplicationCommand
} from './application-commands';

const commands: ApplicationCommand[] = [
  {
    id: '1',
    application_ref: '2@apps.test',
    application_name: 'Poll Bot',
    integration_type: 'guild_install',
    interaction_context: 'guild',
    name: 'poll',
    type: 'chat_input',
    description: 'Create a poll'
  },
  {
    id: '2',
    application_ref: '2@apps.test',
    application_name: 'Poll Bot',
    integration_type: 'guild_install',
    interaction_context: 'guild',
    name: 'about',
    type: 'chat_input'
  }
];

describe('application commands', () => {
  it('keeps remote user-installed commands when guild command use is denied', () => {
    const external = (['chat_input', 'user', 'message'] as const).map((type, index) => ({
      ...commands[0],
      id: `remote-${index}`,
      application_ref: '9@remote.example',
      application_name: 'Remote Tools',
      integration_type: 'user_install' as const,
      type
    }));

    expect(
      [commands[0], ...external].filter((command) =>
        applicationCommandAllowedByUsePermission(command, false)
      )
    ).toEqual(external);
    expect(
      [commands[0], ...external].filter((command) =>
        applicationCommandAllowedByUsePermission(command, true)
      )
    ).toEqual([commands[0], ...external]);
  });

  it('fails closed when a component is missing its installation lineage', () => {
    expect(applicationIntegrationAllowedByUsePermission(null, false)).toBe(false);
    expect(applicationIntegrationAllowedByUsePermission(undefined, false)).toBe(false);
    expect(applicationIntegrationAllowedByUsePermission('user_install', false)).toBe(true);
    expect(applicationIntegrationAllowedByUsePermission('dm_capability', false)).toBe(true);
  });

  it('requires send permission only for user context commands', () => {
    const external = (['chat_input', 'user', 'message'] as const).map((type, index) => ({
      ...commands[0],
      id: `external-${index}`,
      integration_type: 'user_install' as const,
      type
    }));

    expect(
      external.filter((command) =>
        applicationCommandAllowedByChannelPermissions(command, false, false)
      )
    ).toEqual([external[0], external[2]]);
    expect(
      external.filter((command) =>
        applicationCommandAllowedByChannelPermissions(command, false, true)
      )
    ).toEqual(external);
    expect(applicationCommandAllowedByChannelPermissions(commands[0], false, true)).toBe(false);
  });

  it('matches command and application names for composer completion', () => {
    expect(commandCompletions(commands, 'pol').map((item) => item.value)).toEqual([
      '/poll',
      '/about'
    ]);
  });

  it('parses only unambiguous published commands and keeps arguments opaque', () => {
    expect(commandInvocation('/poll lunch tomorrow', commands)).toEqual({
      command: commands[0],
      options: { raw: 'lunch tomorrow' }
    });
    expect(commandInvocation('/missing hello', commands)).toBeNull();
  });

  it('parses lowercase Unicode command names using the Discord naming contract', () => {
    const unicodeCommands: ApplicationCommand[] = [
      { ...commands[0], id: '3', name: 'météo' },
      { ...commands[0], id: '4', name: 'मौसम' },
      { ...commands[0], id: '5', name: 'อากาศ' }
    ];
    expect(commandInvocation('/météo demain', unicodeCommands)?.command.name).toBe('météo');
    expect(commandInvocation('/मौसम', unicodeCommands)?.command.name).toBe('मौसम');
    expect(commandInvocation('/อากาศ', unicodeCommands)?.command.name).toBe('อากาศ');
    expect(commandInvocation('/Météo', unicodeCommands)).toBeNull();
  });

  it('projects localized labels while invoking the default command identity', () => {
    const localized: ApplicationCommand[] = [
      {
        ...commands[0],
        name: 'weather',
        name_localizations: { fr: 'météo', 'en-GB': 'forecast' },
        description: 'Current weather',
        description_localizations: { fr: 'Météo actuelle', 'es-ES': 'Tiempo actual' }
      }
    ];
    expect(commandCompletions(localized, 'mét', 'fr')[0]).toMatchObject({
      value: '/météo',
      label: '/météo',
      detail: 'Météo actuelle · Poll Bot'
    });
    expect(commandInvocation('/météo demain', localized, 'fr')).toEqual({
      command: localized[0],
      options: { raw: 'demain' }
    });
    expect(commandInvocation('/weather tomorrow', localized, 'fr')?.command.name).toBe('weather');
    expect(uniqueChatInputCommand(localized, 'météo', 'fr')?.name).toBe('weather');
    expect(localizedCommandText('Weather', { 'en-GB': 'Forecast' }, 'en-US')).toBe('Forecast');
    expect(localizedCommandText('Weather', { 'es-ES': 'Tiempo' }, 'es-419')).toBe('Tiempo');
  });

  it('does not guess when two commands project the same localized invocation name', () => {
    const ambiguous: ApplicationCommand[] = [
      { ...commands[0], id: '3', name: 'weather', name_localizations: { fr: 'météo' } },
      { ...commands[0], id: '4', name: 'forecast', name_localizations: { fr: 'météo' } }
    ];
    expect(uniqueChatInputCommand(ambiguous, 'météo', 'fr')).toBeNull();
    expect(resolveCommandInvocation('/météo', ambiguous, 'fr')).toEqual({
      kind: 'ambiguous',
      commands: ambiguous
    });
  });

  it('retains exact app and installation identity in same-name completions', () => {
    const duplicate: ApplicationCommand = {
      ...commands[0],
      id: '1',
      application_ref: '9@remote.example',
      application_name: 'Remote Poll',
      integration_type: 'user_install'
    };
    const options = commandCompletions([commands[0], duplicate], 'poll');

    expect(options).toHaveLength(2);
    expect(
      applicationCommandByIdentity([commands[0], duplicate], options[1].applicationCommand)
    ).toBe(duplicate);
    expect(options[1].applicationCommand).toEqual({
      id: '1',
      applicationRef: '9@remote.example',
      integrationType: 'user_install',
      interactionContext: 'guild'
    });
  });

  it('groups the Apps launcher by qualified application before commands', () => {
    const remote = {
      ...commands[0],
      id: '1',
      application_ref: '9@remote.example',
      application_name: 'Remote Poll'
    };
    expect(applicationCommandLauncherGroups([remote, ...commands], 'poll')).toEqual([
      {
        applicationRef: '2@apps.test',
        applicationName: 'Poll Bot',
        commands: [commands[1], commands[0]]
      },
      {
        applicationRef: '9@remote.example',
        applicationName: 'Remote Poll',
        commands: [remote]
      }
    ]);
  });

  it('exposes the bounded string options needed by Discord-style command fields', () => {
    const thread = {
      ...commands[0],
      name: 'thread',
      options: [
        { type: 'string' as const, name: 'name', required: true },
        { type: 'string' as const, name: 'message', required: true },
        { type: 'boolean' as const, name: 'private' }
      ]
    };
    expect(commandStringOptions(thread).map((option) => option.name)).toEqual(['name', 'message']);
    expect(commandOptionsComplete(thread, { name: 'test', message: '' })).toBe(false);
    expect(commandOptionsComplete(thread, { name: 'test', message: 'hello' })).toBe(true);
  });

  it('exposes numeric fields that support dynamic autocomplete', () => {
    const command = {
      ...commands[0],
      options: [
        { type: 'integer' as const, name: 'issue', autocomplete: true },
        { type: 'boolean' as const, name: 'private' }
      ]
    };
    expect(commandComposerOptions(command)).toEqual([
      { type: 'integer', name: 'issue', autocomplete: true },
      { type: 'boolean', name: 'private' }
    ]);
    expect(commandOptionPayload(command, { issue: '42', private: true })).toEqual({
      issue: 42,
      private: true
    });
  });

  it('composes typed options beneath selected subcommand groups', () => {
    const command: ApplicationCommand = {
      ...commands[0],
      name: 'admin',
      options: [
        {
          type: 'subcommand_group',
          name: 'member',
          options: [
            {
              type: 'subcommand',
              name: 'timeout',
              options: [
                { type: 'user', name: 'user', required: true },
                { type: 'integer', name: 'minutes', required: true, min_value: 1, max_value: 60 },
                { type: 'boolean', name: 'notify' }
              ]
            }
          ]
        }
      ]
    };
    const values = {
      '$container:': 'member',
      '$container:member': 'timeout',
      'member.timeout.user': '9@chat.example',
      'member.timeout.minutes': '15',
      'member.timeout.notify': false
    };
    expect(commandOptionsComplete(command, values)).toBe(true);
    expect(commandOptionPayload(command, values)).toEqual({
      member: {
        timeout: {
          user: '9@chat.example',
          minutes: 15,
          notify: false
        }
      }
    });
    expect(commandOptionsComplete(command, { ...values, 'member.timeout.minutes': '90' })).toBe(
      false
    );
  });

  it('returns only attachment uploads consumed by the selected command leaf', () => {
    const command: ApplicationCommand = {
      ...commands[0],
      options: [
        {
          type: 'subcommand',
          name: 'inspect',
          options: [
            { type: 'attachment', name: 'first', required: true },
            { type: 'attachment', name: 'second' },
            { type: 'string', name: 'caption' }
          ]
        },
        {
          type: 'subcommand',
          name: 'other',
          options: [{ type: 'attachment', name: 'file' }]
        }
      ]
    };
    expect(
      commandAttachmentOptionIds(command, {
        '$container:': 'inspect',
        'inspect.first': '91',
        'inspect.second': '92',
        'inspect.caption': 'release',
        'other.file': '93'
      })
    ).toEqual(['91', '92']);
  });

  it('filters channel choices only when the command advertises channel types', () => {
    const unrestricted = { type: 'channel' as const, name: 'destination' };
    const announcements = {
      ...unrestricted,
      channel_types: [5, 10]
    };

    expect(commandOptionAllowsChannelType(unrestricted, 17)).toBe(true);
    expect(commandOptionAllowsChannelType(announcements, 5)).toBe(true);
    expect(commandOptionAllowsChannelType(announcements, 10)).toBe(true);
    expect(commandOptionAllowsChannelType(announcements, 0)).toBe(false);
    expect(commandOptionAllowsChannelType({ type: 'user', name: 'person' }, 5)).toBe(false);
  });

  it('rejects scalar children in federated command tree arrays', () => {
    expect(() =>
      parseApplicationCommands([
        {
          ...commands[0],
          options: [
            { type: 'string', name: 'target', description: 'Target' },
            'silently dropped before this regression'
          ]
        }
      ])
    ).toThrow(/option/u);
    expect(() =>
      parseApplicationCommands([
        {
          ...commands[0],
          options: [
            {
              type: 'string',
              name: 'target',
              description: 'Target',
              choices: [{ name: 'Safe', value: 'safe' }, 7]
            }
          ]
        }
      ])
    ).toThrow(/choice/u);
    expect(() =>
      parseApplicationCommands([
        {
          ...commands[0],
          options: [
            {
              type: 'channel',
              name: 'destination',
              description: 'Destination',
              channel_types: [0, '2']
            }
          ]
        }
      ])
    ).toThrow(/channel types/u);
  });

  it('rejects a malformed autocomplete array instead of partially applying it', () => {
    expect(
      parseApplicationCommandAutocompleteChoices([
        { name: 'Safe', value: 'safe' },
        { name: 'Count', value: 2 }
      ])
    ).toEqual([
      { name: 'Safe', value: 'safe' },
      { name: 'Count', value: 2 }
    ]);
    expect(() =>
      parseApplicationCommandAutocompleteChoices([
        { name: 'Safe', value: 'safe' },
        'silently dropped before this regression'
      ])
    ).toThrow(/autocomplete choice/u);
    expect(() =>
      parseApplicationCommandAutocompleteChoices([{ name: 'Infinite', value: Infinity }])
    ).toThrow(/autocomplete choice/u);
  });

  it('retains exact effective identities and rejects duplicates', () => {
    expect(parseApplicationCommands(commands)).toEqual(commands);
    expect(() => parseApplicationCommands([commands[0], { ...commands[0] }])).toThrow(
      /duplicated/u
    );
  });

  it('retains bot-DM capability lineage in plaintext and encrypted requests', () => {
    const capability = parseApplicationCommands([
      {
        ...commands[0],
        integration_type: 'dm_capability',
        interaction_context: 'bot_dm',
        dm_capability_id: `kbdg_${'a'.repeat(43)}`,
        dm_capability_revision: '7'
      }
    ])[0];

    expect(applicationCommandRequestIdentity(capability)).toEqual({
      application_ref: '2@apps.test',
      command_id: '1',
      integration_type: 'dm_capability',
      command_name: 'poll',
      command_type: 'chat_input',
      dm_capability_id: `kbdg_${'a'.repeat(43)}`,
      dm_capability_revision: '7'
    });
    expect(() =>
      parseApplicationCommands([
        {
          ...commands[0],
          integration_type: 'dm_capability',
          interaction_context: 'bot_dm'
        }
      ])
    ).toThrow(/capability lineage/u);
  });
});

File: mobile/test/application_commands_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:kaede_mobile/src/core/refs.dart';
import 'package:kaede_mobile/src/domain/application_commands.dart';

void main() {
  final application = EntityRef.parse('7@apps.example');
  MobileApplicationCommand command(String name, String type) =>
      MobileApplicationCommand(
        id: name,
        application: application,
        applicationName: 'Tools',
        name: name,
        type: type,
        description: '',
        integrationType: 'user_install',
        interactionContext: 'private_channel',
      );

  test('guild denial keeps remote user-installed slash and context commands',
      () {
    final guild = MobileApplicationCommand(
      id: 'guild',
      application: application,
      applicationName: 'Guild tools',
      name: 'guild',
      type: 'chat_input',
      description: '',
      integrationType: 'guild_install',
      interactionContext: 'guild',
    );
    final external = <MobileApplicationCommand>[
      for (final type in const <String>['chat_input', 'user', 'message'])
        MobileApplicationCommand(
          id: 'remote-$type',
          application: EntityRef.parse('8@remote.example'),
          applicationName: 'Remote tools',
          name: 'remote-$type',
          type: type,
          description: '',
          integrationType: 'user_install',
          interactionContext: 'guild',
        ),
    ];

    expect(
      <MobileApplicationCommand>[guild, ...external].where((item) =>
          mobileApplicationCommandAllowedByUsePermission(item, false)),
      external,
    );
    expect(
      <MobileApplicationCommand>[guild, ...external].where(
          (item) => mobileApplicationCommandAllowedByUsePermission(item, true)),
      <MobileApplicationCommand>[guild, ...external],
    );
  });

  test('component permission checks fail closed without installation lineage',
      () {
    expect(
      mobileApplicationIntegrationAllowedByUsePermission(null, false),
      isFalse,
    );
    expect(
      mobileApplicationIntegrationAllowedByUsePermission('unknown', false),
      isFalse,
    );
    expect(
      mobileApplicationIntegrationAllowedByUsePermission('user_install', false),
      isTrue,
    );
    expect(
      mobileApplicationIntegrationAllowedByUsePermission(
          'dm_capability', false),
      isTrue,
    );
  });

  test('send permission is required only for user context commands', () {
    final external = <MobileApplicationCommand>[
      command('slash', 'chat_input'),
      command('user', 'user'),
      command('message', 'message'),
    ];

    expect(
      external.where((item) =>
          mobileApplicationCommandAllowedByChannelPermissions(
              item, false, false)),
      <MobileApplicationCommand>[external[0], external[2]],
    );
    expect(
      external.where((item) =>
          mobileApplicationCommandAllowedByChannelPermissions(
              item, false, true)),
      external,
    );
  });

  test('Apps menus expose only context commands for their target surface', () {
    final user = command('Inspect user', 'user');
    final message = command('Quote message', 'message');
    final slash = command('search', 'chat_input');

    expect(mobileUserContextCommands([user, message, slash]), [user]);
    expect(
        mobileMessageContextCommands([user, message, slash]), [user, message]);
  });

  test('context Apps groups are app-first and searchable', () {
    final local = command('Inspect user', 'user');
    final remote = MobileApplicationCommand(
      id: 'remote-review',
      application: EntityRef.parse('8@remote.example'),
      applicationName: 'Moderation',
      name: 'review',
      nameLocalizations: const <String, String>{'es-ES': 'revisar'},
      type: 'message',
      description: 'Review a message',
      integrationType: 'guild_install',
      interactionContext: 'guild',
    );

    expect(
      mobileContextCommandGroups([local, remote], '', 'en-US')
          .map((group) => group.applicationName),
      ['Moderation', 'Tools'],
    );
    expect(
      mobileContextCommandGroups([local, remote], 'moderat', 'en-US')
          .single
          .commands,
      [remote],
    );
    expect(
      mobileContextCommandGroups([local, remote], 'revisar', 'es-ES')
          .single
          .commands,
      [remote],
    );
    expect(
      mobileContextCommandGroups([local, remote], 'missing', 'en-US'),
      isEmpty,
    );
  });

  test('context Apps keeps all fifteen commands of each supported type', () {
    final commands = <MobileApplicationCommand>[
      for (var index = 0; index < 15; index++)
        command('message-$index', 'message'),
      for (var index = 0; index < 15; index++) command('user-$index', 'user'),
    ];

    expect(
      mobileContextCommandGroups(commands, '', 'en-US').single.commands,
      hasLength(30),
    );
    expect(
      mobileContextCommandGroups(commands, 'message-', 'en-US').single.commands,
      hasLength(15),
    );
    expect(
      mobileContextCommandGroups(commands, 'user-', 'en-US').single.commands,
      hasLength(15),
    );
  });

  test('context Apps hoists bounded successful history per account', () {
    final inspect = command('inspect', 'user');
    final quote = command('quote', 'message');
    var history = <String>[];
    history = mobileRememberContextCommand(history, quote);
    history = mobileRememberContextCommand(history, inspect);
    history = mobileRememberContextCommand(history, inspect);

    final model = mobileContextCommandMenuModel(
      [quote, inspect],
      '',
      'en-US',
      history,
    );
    expect(model.frequent, [inspect, quote]);
    expect(model.groups, isEmpty);
    expect(
      mobileContextCommandHistoryStorageKey(
        EntityRef.parse('7@users.example'),
      ),
      isNot(
        mobileContextCommandHistoryStorageKey(
          EntityRef.parse('8@users.example'),
        ),
      ),
    );

    for (var index = 0; index < 110; index += 1) {
      history = mobileRememberContextCommand(history, quote);
    }
    expect(history, hasLength(100));
    expect(history.last, mobileContextCommandUsageKey(quote));
  });

  test('context commands bind the exact user or message identity', () {
    final user = EntityRef.parse('8@users.example');
    final message = EntityRef.parse('9@chat.example');

    expect(
      mobileContextCommandTarget(command('Inspect user', 'user'), user: user),
      user,
    );
    expect(
      mobileContextCommandTarget(
        command('Quote message', 'message'),
        user: user,
        message: message,
      ),
      message,
    );
    expect(
      () => mobileContextCommandTarget(
        command('Quote message', 'message'),
        user: user,
      ),
      throwsA(isA<FormatException>()),
    );
  });

  test('authority-selected user install and DM context remain typed', () {
    final parsed = MobileApplicationCommand.fromJson(<String, Object?>{
      'application_ref': '7@apps.example',
      'id': '91',
      'application_name': 'Tools',
      'integration_type': 'user_install',
      'interaction_context': 'bot_dm',
      'name': 'inspect',
      'type': 'chat_input',
      'description': 'Inspect a record',
    });

    expect(parsed.integrationType, 'user_install');
    expect(parsed.interactionContext, 'bot_dm');
    expect(parsed.id, '91');
  });

  test('bot-DM discovery retains exact capability lineage', () {
    final parsed = MobileApplicationCommand.fromJson(<String, Object?>{
      'application_ref': '7@apps.example',
      'id': '91',
      'application_name': 'Tools',
      'integration_type': 'dm_capability',
      'dm_capability_id': 'kbdg_${List.filled(43, 'a').join()}',
      'dm_capability_revision': '7',
      'interaction_context': 'bot_dm',
      'name': 'inspect',
      'type': 'chat_input',
      'description': 'Inspect a record',
    });

    expect(parsed.integrationType, 'dm_capability');
    expect(parsed.dmCapabilityId, 'kbdg_${List.filled(43, 'a').join()}');
    expect(parsed.dmCapabilityRevision, '7');
    expect(
      () => MobileApplicationCommand.fromJson(<String, Object?>{
        'application_ref': '7@apps.example',
        'id': '91',
        'application_name': 'Tools',
        'integration_type': 'dm_capability',
        'interaction_context': 'bot_dm',
        'name': 'inspect',
        'type': 'chat_input',
        'description': 'Inspect a record',
      }),
      throwsA(isA<FormatException>()),
    );
  });

  test('same-name commands retain stable app identity and remain ambiguous',
      () {
    final local = command('inspect', 'chat_input');
    final remote = MobileApplicationCommand(
      id: 'inspect-remote',
      application: EntityRef.parse('8@remote.example'),
      applicationName: 'Remote tools',
      name: 'inspect',
      type: 'chat_input',
      description: 'Inspect remotely',
      integrationType: 'guild_install',
      interactionContext: 'guild',
    );

    expect(mobileChatInputCommandMatches([local, remote], 'inspect', 'en-US'),
        [local, remote]);
    final groups = mobileApplicationCommandLauncherGroups(
      [remote, local],
      'inspect',
      'en-US',
    );
    expect(groups.map((group) => group.application),
        [EntityRef.parse('8@remote.example'), application]);
  });

  test('attachment file types are parsed and matched before upload', () {
    final parsed = MobileApplicationCommandOption.fromJson(<String, Object?>{
      'name': 'evidence',
      'type': 'attachment',
      'description': 'Evidence',
      'file_types': <String>['IMAGE', '.PDF'],
    });

    expect(parsed.fileTypes, ['image', '.pdf']);
    expect(mobileCommandFileMatches(parsed, 'photo.png', 'image/png'), isTrue);
    expect(mobileCommandFileMatches(parsed, 'report.PDF', 'application/pdf'),
        isTrue);
    expect(
        mobileCommandFileMatches(
            parsed, 'payload.exe', 'application/octet-stream'),
        isFalse);
  });

  test('federated command trees reject scalar array children', () {
    Map<String, Object?> commandJson(Object options) => <String, Object?>{
          'application_ref': '7@apps.example',
          'id': '91',
          'application_name': 'Tools',
          'integration_type': 'user_install',
          'interaction_context': 'private_channel',
          'name': 'inspect',
          'type': 'chat_input',
          'description': 'Inspect a record',
          'options': options,
        };

    expect(
      () => MobileApplicationCommand.fromJson(commandJson(<Object>[
        <String, Object?>{
          'name': 'target',
          'type': 'string',
          'description': 'Target',
        },
        'silently dropped before this regression',
      ])),
      throwsA(isA<FormatException>()),
    );
    expect(
      () => MobileApplicationCommand.fromJson(commandJson(<Object>[
        <String, Object?>{
          'name': 'target',
          'type': 'string',
          'description': 'Target',
          'choices': <Object>[
            <String, Object?>{'name': 'Safe', 'value': 'safe'},
            7,
          ],
        },
      ])),
      throwsA(isA<FormatException>()),
    );
    expect(
      () => MobileApplicationCommand.fromJson(commandJson(<Object>[
        <String, Object?>{
          'name': 'channel',
          'type': 'channel',
          'description': 'Channel',
          'channel_types': <Object>[0, '2'],
        },
      ])),
      throwsA(isA<FormatException>()),
    );
  });

  test(
      'autocomplete rejects a malformed array instead of partially applying it',
      () {
    expect(
      mobileAutocompleteChoices(<Object?>[
        <String, Object?>{'name': 'Safe', 'value': 'safe'},
        <String, Object?>{'name': 'Count', 'value': 2},
      ]).map((choice) => choice.value),
      <Object>['safe', 2],
    );
    expect(
      () => mobileAutocompleteChoices(<Object?>[
        <String, Object?>{'name': 'Safe', 'value': 'safe'},
        'silently dropped before this regression',
      ]),
      throwsA(isA<FormatException>()),
    );
    expect(
      () => mobileAutocompleteChoices(<Object?>[
        <String, Object?>{'name': 'Infinite', 'value': double.infinity},
      ]),
      throwsA(isA<FormatException>()),
    );
  });
}
